### PR TITLE
executor: snapshot for (batch) point get not working correctly

### LINKED ISSUE #25167: stale read like 'select ... as of timestamp ...' for point get not work'
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table test.table1 (id int primary key, a int);
insert into test.table1 values(1, 1);
select sleep(2);update test.table1 set a=2 where id=1;select * from test.table1 as of timestamp TIMESTAMP(NOW() - INTERVAL 1 SECOND) where id=1;
```

### 2. What did you expect to see? (Required)

```
+----+------+
| id | a    |
+----+------+
|  1 |    1 |
+----+------+
```

### 3. What did you see instead (Required)

```
+----+------+
| id | a    |
+----+------+
|  1 |    2 |
+----+------+
```

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by xhebox**
Signed-off-by: xhe <xw897002528@gmail.com>

### What problem does this PR solve?

Issue Number: close #25167 

Problem Summary: We may request an old snapshot, where the cached `txnCtx` may not provide.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- Fix stale reads and history reads for (batch) point get

### CODE REVIEW

**Comment by bb7133** [comment]
Please add a test case for your change.

**Review by bb7133 — COMMENTED** [review]
Please add a test case for your change.

**Comment by xhebox** [comment]
> Please add a test case for your change.

I don't think it is needed, but here you go.

File: executor/batch_point_get_test.go [hunk 1]
```diff
@@ -346,3 +346,16 @@ func (s *testBatchPointGetSuite) TestPointGetForTemporaryTable(c *C) {
 	tk.MustQuery("select * from t1 where id = 1").Check(testkit.Rows("1 1"))
 	tk.MustQuery("select * from t1 where id = 2").Check(testkit.Rows())
 }
+
+func (s *testBatchPointGetSuite) TestBatchPointGetIssue25167(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a int)")
```
Comment (nolouch) [inline executor/batch_point_get_test.go:354]: should be
```suggestion
	tk.MustExec("create table t (a int) primary key")
```

File: executor/point_get_test.go [hunk 2]
```diff
@@ -944,3 +944,16 @@ func (s *testPointGetSuite) TestWithTiDBSnapshot(c *C) {
 
 	tk.MustQuery("select * from xx").Check(testkit.Rows("1", "7"))
 }
+
+func (s *testPointGetSuite) TestPointGetIssue25167(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a int)")
```
Comment (nolouch) [inline executor/point_get_test.go:952]:
```suggestion
	tk.MustExec("create table t (a int) primary key")
```

**Review by nolouch — COMMENTED** [review]
rest LGTM

File: executor/batch_point_get_test.go (hunk 1 above)
Comment (xhebox) [reply executor/batch_point_get_test.go:354]: It is not related to the root cause, though, whatever.

**Review by nolouch — APPROVED** [review]
lgtm

**Review by JmPotato — APPROVED** [review]

**Review by bb7133 — APPROVED** [review]

**Pull request merged**